### PR TITLE
Avoid the Caffeine overhead when it's not used

### PR DESCRIPTION
**Pull request by Sanne**

### CODE REVIEW

File: extensions/caffeine/runtime/src/main/java/io/quarkus/caffeine/runtime/graal/CacheConstructorsAutofeature.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+                "com.github.benmanes.caffeine.cache.PSWMW",
+                "com.github.benmanes.caffeine.cache.SILMS",
+                "com.github.benmanes.caffeine.cache.SSA",
+                "com.github.benmanes.caffeine.cache.SSLA",
+                "com.github.benmanes.caffeine.cache.SSLMS",
+                "com.github.benmanes.caffeine.cache.SSMS",
+                "com.github.benmanes.caffeine.cache.SSMSA",
+                "com.github.benmanes.caffeine.cache.SSMSW",
+                "com.github.benmanes.caffeine.cache.SSW",
+        };
```
Comment (Sanne) [inline extensions/caffeine/runtime/src/main/java/io/quarkus/caffeine/runtime/graal/CacheConstructorsAutofeature.java:85]: N.B. the list has grown slightly, as I've merged the missing ones from the (now deleted) duplicate registrations that were being triggered by https://github.com/quarkusio/quarkus/pull/17748/files#diff-b8f55d2a209ec481eaa58dee37f596a382aa4443a9585ae00581fcfac9a9c8b3

File: extensions/caffeine/runtime/src/main/java/io/quarkus/caffeine/runtime/graal/CacheConstructorsAutofeature.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+ */
+@AutomaticFeature
+public class CacheConstructorsAutofeature implements Feature {
+
+    private final AtomicBoolean triggered = new AtomicBoolean(false);
+
+    /**
+     * To set this, add `-J-Dorg.quarkus.caffeine.graalvm.diagnostics=true` to the native-image parameters
+     */
+    private static final boolean log = Boolean.getBoolean("org.quarkus.caffeine.graalvm.diagnostics");
```
Comment (gsmet) [inline extensions/caffeine/runtime/src/main/java/io/quarkus/caffeine/runtime/graal/CacheConstructorsAutofeature.java:31]:
```suggestion
    private static final boolean log = Boolean.getBoolean("io.quarkus.caffeine.graalvm.diagnostics");
```

File: extensions/caffeine/runtime/src/main/java/io/quarkus/caffeine/runtime/graal/CacheConstructorsAutofeature.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+ * Caffeine is indeed being used: only if the cache builder is reachable
+ * in the application code.
+ */
+@AutomaticFeature
+public class CacheConstructorsAutofeature implements Feature {
+
+    private final AtomicBoolean triggered = new AtomicBoolean(false);
+
+    /**
+     * To set this, add `-J-Dorg.quarkus.caffeine.graalvm.diagnostics=true` to the native-image parameters
```
Comment (gsmet) [inline extensions/caffeine/runtime/src/main/java/io/quarkus/caffeine/runtime/graal/CacheConstructorsAutofeature.java:29]:
```suggestion
     * To set this, add `-J-Dio.quarkus.caffeine.graalvm.diagnostics=true` to the native-image parameters
```

**Review by gsmet — CHANGES_REQUESTED** [review]
Interesting. I'm wondering if we could be even more clever and register the classes depending on the options used in the DSL. But given how many combinations there are, it would probably be impractical.

Added a small comment about the groupId/domain name we normally use.

**Comment by Sanne** [comment]
> Interesting. I'm wondering if we could be even more clever and register the classes depending on the options used in the DSL. But given how many combinations there are, it would probably be impractical.

+1 There are likely further ways to pre-filter this list.

I'm not sure if I can look at that immediately though, I'd suggest to merge this and then we can see what else is possible.

**Comment by Sanne** [comment]
rebased to apply your suggestions on clean history

**Comment by Sanne** [comment]
right looks like I can easily split the list in two... I'll do it soon.

**Comment by Sanne** [comment]
@gsmet woot, all green on CI

**Review by gsmet — APPROVED** [review]

**Pull request merged**